consensus/reactor.go

```diff
@@ -760,967 +662,720 @@ func (conR *Reactor) gossipVotesForHeight(
-	ErrPeerStateInvalidStartTime = errors.New("error peer state invalid startTime")
-)
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Do not allow starting new broadcasting goroutines after reactor shutdown
+		// has been initiated. This can happen after we've manually closed all
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
+			return
+		}
 
-// PeerState contains the known state of a peer, including its connection and
-// threadsafe access to its PeerRoundState.
-// NOTE: THIS GETS DUMPED WITH rpc/core/consensus.go.
-// Be mindful of what you Expose.
-type PeerState struct {
-	peer   p2p.Peer
-	logger log.Logger
-
-	mtx   sync.Mutex             // NOTE: Modify below using setters, never directly.
-	PRS   cstypes.PeerRoundState `json:"round_state"` // Exposed.
-	Stats *peerStateStats        `json:"stats"`       // Exposed.
-}
+		var (
+			ps *PeerState
+			ok bool
+		)
 
-// peerStateStats holds internal statistics for a peer.
-type peerStateStats struct {
-	Votes      int `json:"votes"`
-	BlockParts int `json:"block_parts"`
-}
+		ps, ok = r.peers[peerUpdate.PeerID]
+		if !ok {
+			ps = NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers[peerUpdate.PeerID] = ps
+		}
 
-func (pss peerStateStats) String() string {
-	return fmt.Sprintf("peerStateStats{votes: %d, blockParts: %d}",
-		pss.Votes, pss.BlockParts)
-}
+		if !ps.IsRunning() {
+			// Set the peer state's closer to signal to all spawned goroutines to exit
+			// when the peer is removed. We also set the running state to ensure we
+			// do not spawn multiple instances of the same goroutines and finally we
+			// set the waitgroup counter so we know when all goroutines have exited.
+			ps.broadcastWG.Add(3)
+			ps.SetRunning(true)
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(ps)
+			go r.gossipVotesRoutine(ps)
+			go r.queryMaj23Routine(ps)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.waitSync {
+				r.sendNewRoundStepMessage(ps.peerID)
+			}
+		}
 
-// NewPeerState returns a new PeerState for the given Peer
-func NewPeerState(peer p2p.Peer) *PeerState {
-	return &PeerState{
-		peer:   peer,
-		logger: log.NewNopLogger(),
-		PRS: cstypes.PeerRoundState{
-			Round:              -1,
-			ProposalPOLRound:   -1,
-			LastCommitRound:    -1,
-			CatchupCommitRound: -1,
-		},
-		Stats: &peerStateStats{},
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		ps, ok := r.peers[peerUpdate.PeerID]
+		if ok && ps.IsRunning() {
+			// signal to all spawned goroutines for the peer to gracefully exit
+			ps.closer.Close()
+
+			go func() {
+				// Wait for all spawned broadcast goroutines to exit before marking the
+				// peer state as no longer running and removal from the peers map.
+				ps.broadcastWG.Wait()
+				ps.SetRunning(false)
```

Comment:
We should delete from the map before setting running to false, otherwise there is a race where e.g. we set running=false, the main thread picks the peer state from the map, sees that it is not running, and starts it again.